Purchases/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -302,22 +305,58 @@ class PurchasesOrchestrator {
     }
 
     @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
-    func purchase(sk2Product product: SK2Product, completion: @escaping PurchaseCompletedBlock) {
+    func purchase(sk2Product product: SK2Product,
+                  discount: StoreProductDiscountType?,
+                  completion: @escaping PurchaseCompletedBlock) {
         _ = Task<Void, Never> {
-            let result = await purchase(sk2Product: product)
-            DispatchQueue.main.async {
-                switch result {
-                case .failure(let error) where error is StoreKitError:
-                    completion(nil, nil, ErrorUtils.purchasesError(withStoreKitError: error), false)
-                case .failure(let error):
+            do {
+                let result: PurchaseResultData = try await self.purchase(sk2Product: product, discount: discount)
+
+                DispatchQueue.main.async {
+                    completion(result.0, result.1, nil, result.2)
+                }
+            } catch let error {
+                DispatchQueue.main.async {
                     completion(nil, nil, error, false)
-                case .success(let (customerInfo, userCancelled)):
-                    // todo: change API and send transaction
-                    if userCancelled {
-                        completion(nil, nil, ErrorUtils.purchaseCancelledError(), userCancelled)
```

Comment:
This lead to a bug: cancelling was actually `throw`ing, which is not what the documentation said.